src/math.rs

```diff
@@ -0,0 +1,132 @@
+use rand::Rng as _;
+use rand::distributions::Distribution as _;
+use rustc_apfloat::Float as _;
+use rustc_apfloat::ieee::IeeeFloat;
+
+/// Disturbes a floating-point result by a relative error on the order of (-2^scale, 2^scale).
+pub(crate) fn apply_random_float_error<F: rustc_apfloat::Float>(
+    ecx: &mut crate::MiriInterpCx<'_>,
+    val: F,
+    err_scale: i32,
+) -> F {
+    let rng = ecx.machine.rng.get_mut();
+    // Generate a random integer in the range [0, 2^PREC).
+    let dist = rand::distributions::Uniform::new(0, 1 << F::PRECISION);
+    let err = F::from_u128(dist.sample(rng))
+        .value
+        .scalbn(err_scale.strict_sub(F::PRECISION.try_into().unwrap()));
+    // give it a random sign
+    let err = if rng.gen::<bool>() { -err } else { err };
+    // multiple the value with (1+err)
+    (val * (F::from_u128(1).value + err).value).value
+}
+
+pub(crate) fn sqrt<S: rustc_apfloat::ieee::Semantics>(x: IeeeFloat<S>) -> IeeeFloat<S> {
+    match x.category() {
+        // preserve zero sign
+        rustc_apfloat::Category::Zero => x,
+        // propagate NaN
+        rustc_apfloat::Category::NaN => x,
+        // sqrt of negative number is NaN
+        _ if x.is_negative() => IeeeFloat::NAN,
+        // sqrt(∞) = ∞
+        rustc_apfloat::Category::Infinity => IeeeFloat::INFINITY,
+        rustc_apfloat::Category::Normal => {
+            // Floating point precision, excluding the integer bit
+            let prec = i32::try_from(S::PRECISION).unwrap() - 1;
+
+            // x = 2^(exp - prec) * mant
+            let mut exp = x.ilogb();
+            let mut mant = x.scalbn(prec - exp).to_u128(128).value;
+
+            if exp % 2 != 0 {
+                // Make exponent even, so it can be divided by 2
+                exp -= 1;
+                mant <<= 1;
+            }
+
+            // Bit-by-bit sqrt with one extra bit
+            let mut rem = mant << 1; // extra bit
+            let mut res = 0u128;
+            let mut s = 0u128;
+            let mut d = 1u128 << (prec + 1);
+
+            while d != 0 {
+                let t = s + d;
+                if rem >= t {
+                    s += d + d;
+                    rem -= t;
+                    res += d;
+                }
+                rem <<= 1;
+                d >>= 1;
+            }
+
+            // Round to nearest and remove extra bit
+            res = (res + (res & 1)) >> 1;
```

Comment:
Round to nearest what? Is this the same as `res.next_multiple_of(2) >> 1`?

Reply:
So, is this the same as `res.next_multiple_of(2) >> 1`?

When a number is odd, it is equally near to both it even neighbor, so I don't get the "rounding to nearest" part.

Reply:
It is the same as `next_multiple_of`, but I think the inline expression better expresses the intent of "isolate the lowest bit and add it to the value".

When the lowest bit is 1, the result is irrational (otherwise, the lowest ~prec/2 bits would be zero) and there is an infinite amount of 1 bits that have been truncated, so it is actually closer to the upper neighbor.

Reply:
> but I think the inline expression better expresses the intent of "isolate the lowest bit and add it to the value".

But it doesn't explain what "isolate the lowest bit and add it to the value" has to do with "round to nearest".

> When the lowest bit is 1, the result is irrational (otherwise, the lowest ~prec/2 bits would be zero) and there is an infinite amount of 1 bits that have been truncated, so it (the full irrational result) is actually closer to the upper neighbor.

Okay I am completely lost now.^^ Can you expand this explanation (why are there infinitely many 1) and add it as a comment in the code?

Reply:
I don't know if the comment I added makes it clearer.

Reply:
Hm, not quite.

Let me try to re-phrase the comment in my own words. Is this correct?
```
            // Remove extra fractional bit from result, rounding to nearest.
            // If the last bit is 0, then the nearest neighbor is definitely the lower one.
            // If the last bit is 1, it sounds like this may either be a tie (if there's infinitely many
            // 0s after this 1), or the nearest neighbor is the upper one. However,
            // since square roots are either exact or irrational, and an exact root would
            // lead to the last "extra" bit being 0, we can exclude a tie in this case.
            // We therefore always round up if the last bit is 1.
```

Reply:
Okay, thanks for explaining and checking! Please put this comment in the code. :)

Reply:
Done :)